# Data Analysis Notebook
Importación de librerías


In [78]:
import pandas as pd
import matplotlib
import sklearn
from sklearn import preprocessing, neighbors, metrics, pipeline
from sklearn.pipeline import Pipeline
from sklearn.model_selection import TimeSeriesSplit, cross_val_score

Lectura de ficheros de datos en formato csv

In [79]:
wind_ava=pd.read_csv('wind_ava.csv.gz', compression = "gzip")

wind_comp=pd.read_csv('wind_comp.csv.gz', compression = "gzip")

Visualización de las primeras muestras de datos de los ficheros iniciales

In [80]:
wind_ava.head()

datetime   energy     p54.162.1     p54.162.2     p54.162.3  \
0  2005-01-02 18:00:00   402.71  2.534970e+06  2.526864e+06  2.518754e+06   
1  2005-01-03 00:00:00   696.80  2.537369e+06  2.529277e+06  2.521184e+06   
2  2005-01-03 06:00:00  1591.15  2.533727e+06  2.525703e+06  2.517678e+06   
3  2005-01-03 12:00:00  1338.62  2.534491e+06  2.526548e+06  2.518609e+06   
4  2005-01-03 18:00:00   562.50  2.529543e+06  2.521623e+06  2.513702e+06   

      p54.162.4     p54.162.5     p54.162.6     p54.162.7     p54.162.8  ...  \
0  2.510648e+06  2.502537e+06  2.531111e+06  2.522721e+06  2.514330e+06  ...   
1  2.513088e+06  2.504995e+06  2.533465e+06  2.525088e+06  2.516716e+06  ...   
2  2.509654e+06  2.501629e+06  2.529801e+06  2.521496e+06  2.513187e+06  ...   
3  2.510670e+06  2.502732e+06  2.530569e+06  2.522346e+06  2.514127e+06  ...   
4  2.505782e+06  2.497861e+06  2.525621e+06  2.517421e+06  2.509215e+06  ...   

    v100.16   v100.17   v100.18   v100.19   v100.20   v100.21   v100.22  \
0 -4.683596 -4.545396 -4.407196 -4.268996 -4.131295 -4.669626 -4.528932   
1 -3.397886 -3.257192 -3.115998 -2.975304 -2.834609 -3.396390 -3.254198   
2 -1.454105 -1.296447 -1.138290 -0.980134 -0.822476 -1.459094 -1.302933   
3  1.255015  1.370265  1.485515  1.600765  1.716015  1.210612  1.319376   
4  1.939031  2.023847  2.108663  2.193977  2.278793  1.873673  1.953000   

    v100.23   v100.24   v100.25  
0 -4.388736 -4.248540 -4.107846  
1 -3.112506 -2.970314 -2.828622  
2 -1.147271 -0.991110 -0.834949  
3  1.428140  1.536405  1.645169  
4  2.031829  2.111157  2.189986  

[5 rows x 552 columns]

In [81]:
wind_ava.head()

datetime   energy     p54.162.1     p54.162.2     p54.162.3  \
0  2005-01-02 18:00:00   402.71  2.534970e+06  2.526864e+06  2.518754e+06   
1  2005-01-03 00:00:00   696.80  2.537369e+06  2.529277e+06  2.521184e+06   
2  2005-01-03 06:00:00  1591.15  2.533727e+06  2.525703e+06  2.517678e+06   
3  2005-01-03 12:00:00  1338.62  2.534491e+06  2.526548e+06  2.518609e+06   
4  2005-01-03 18:00:00   562.50  2.529543e+06  2.521623e+06  2.513702e+06   

      p54.162.4     p54.162.5     p54.162.6     p54.162.7     p54.162.8  ...  \
0  2.510648e+06  2.502537e+06  2.531111e+06  2.522721e+06  2.514330e+06  ...   
1  2.513088e+06  2.504995e+06  2.533465e+06  2.525088e+06  2.516716e+06  ...   
2  2.509654e+06  2.501629e+06  2.529801e+06  2.521496e+06  2.513187e+06  ...   
3  2.510670e+06  2.502732e+06  2.530569e+06  2.522346e+06  2.514127e+06  ...   
4  2.505782e+06  2.497861e+06  2.525621e+06  2.517421e+06  2.509215e+06  ...   

    v100.16   v100.17   v100.18   v100.19   v100.20   v100.21   v100.22  \
0 -4.683596 -4.545396 -4.407196 -4.268996 -4.131295 -4.669626 -4.528932   
1 -3.397886 -3.257192 -3.115998 -2.975304 -2.834609 -3.396390 -3.254198   
2 -1.454105 -1.296447 -1.138290 -0.980134 -0.822476 -1.459094 -1.302933   
3  1.255015  1.370265  1.485515  1.600765  1.716015  1.210612  1.319376   
4  1.939031  2.023847  2.108663  2.193977  2.278793  1.873673  1.953000   

    v100.23   v100.24   v100.25  
0 -4.388736 -4.248540 -4.107846  
1 -3.112506 -2.970314 -2.828622  
2 -1.147271 -0.991110 -0.834949  
3  1.428140  1.536405  1.645169  
4  2.031829  2.111157  2.189986  

[5 rows x 552 columns]

Si modificamos el índice para que este sea la variable *datetime* nos será más simple operar con ello.

In [82]:
wind_ava = wind_ava.set_index('datetime')
wind_ava.head()

energy     p54.162.1     p54.162.2     p54.162.3  \
datetime                                                                 
2005-01-02 18:00:00   402.71  2.534970e+06  2.526864e+06  2.518754e+06   
2005-01-03 00:00:00   696.80  2.537369e+06  2.529277e+06  2.521184e+06   
2005-01-03 06:00:00  1591.15  2.533727e+06  2.525703e+06  2.517678e+06   
2005-01-03 12:00:00  1338.62  2.534491e+06  2.526548e+06  2.518609e+06   
2005-01-03 18:00:00   562.50  2.529543e+06  2.521623e+06  2.513702e+06   

                        p54.162.4     p54.162.5     p54.162.6     p54.162.7  \
datetime                                                                      
2005-01-02 18:00:00  2.510648e+06  2.502537e+06  2.531111e+06  2.522721e+06   
2005-01-03 00:00:00  2.513088e+06  2.504995e+06  2.533465e+06  2.525088e+06   
2005-01-03 06:00:00  2.509654e+06  2.501629e+06  2.529801e+06  2.521496e+06   
2005-01-03 12:00:00  2.510670e+06  2.502732e+06  2.530569e+06  2.522346e+06   
2005-01-03 18:00:00  2.505782e+06  2.497861e+06  2.525621e+06  2.517421e+06   

                        p54.162.8     p54.162.9  ...   v100.16   v100.17  \
datetime                                         ...                       
2005-01-02 18:00:00  2.514330e+06  2.505940e+06  ... -4.683596 -4.545396   
2005-01-03 00:00:00  2.516716e+06  2.508339e+06  ... -3.397886 -3.257192   
2005-01-03 06:00:00  2.513187e+06  2.504882e+06  ... -1.454105 -1.296447   
2005-01-03 12:00:00  2.514127e+06  2.505904e+06  ...  1.255015  1.370265   
2005-01-03 18:00:00  2.509215e+06  2.501015e+06  ...  1.939031  2.023847   

                      v100.18   v100.19   v100.20   v100.21   v100.22  \
datetime                                                                
2005-01-02 18:00:00 -4.407196 -4.268996 -4.131295 -4.669626 -4.528932   
2005-01-03 00:00:00 -3.115998 -2.975304 -2.834609 -3.396390 -3.254198   
2005-01-03 06:00:00 -1.138290 -0.980134 -0.822476 -1.459094 -1.302933   
2005-01-03 12:00:00  1.485515  1.600765  1.716015  1.210612  1.319376   
2005-01-03 18:00:00  2.108663  2.193977  2.278793  1.873673  1.953000   

                      v100.23   v100.24   v100.25  
datetime                                           
2005-01-02 18:00:00 -4.388736 -4.248540 -4.107846  
2005-01-03 00:00:00 -3.112506 -2.970314 -2.828622  
2005-01-03 06:00:00 -1.147271 -0.991110 -0.834949  
2005-01-03 12:00:00  1.428140  1.536405  1.645169  
2005-01-03 18:00:00  2.031829  2.111157  2.189986  

[5 rows x 551 columns]


Actualmente, el fichero tiene 551 columnas. No las necesitamos todas: sólo usaremos la información sobre la localización de Sotavento, es decir, el número 13 de la cuadrícula de 5x5. 

A continuación, modificamos el fichero de datos para eliminar las columnas que no necesitamos y quedarnos únicamente con las columnas de variables relativas a la posición 13 de la cuadrícula. Para ello, filtramos las columnas según aquellas que contengan el número 13 en su nombre.
Además, nos interesa saber el valor de la energía en cada momento en esa localización, por lo que también filtraremos de forma que no se elimine la variable 'energy'.

In [83]:
for c in wind_ava.columns:
    if not c.endswith('.13') and c != 'energy' and c != 'datetime':
        wind_ava = wind_ava.drop(c, axis = 1)

In [84]:
wind_ava.head()

energy    p54.162.13  p55.162.13    cape.13  \
datetime                                                            
2005-01-02 18:00:00   402.71  2.510824e+06    9.186295  13.527577   
2005-01-03 00:00:00   696.80  2.513173e+06    8.849569   6.896412   
2005-01-03 06:00:00  1591.15  2.509627e+06    7.924080   4.774439   
2005-01-03 12:00:00  1338.62  2.510571e+06    6.922709   0.000000   
2005-01-03 18:00:00   562.50  2.505664e+06    6.646282   0.000000   

                       p59.162.13  lai_lv.13  lai_hv.13   u10n.13   v10n.13  \
datetime                                                                      
2005-01-02 18:00:00  1.386937e+06   2.344111   2.432983 -0.757587 -1.922799   
2005-01-03 00:00:00  1.153526e+06   2.343719   2.432838 -1.412620 -1.403011   
2005-01-03 06:00:00  1.098754e+06   2.343300   2.432704 -2.290185 -0.754580   
2005-01-03 12:00:00  1.076021e+06   2.342830   2.432514 -3.497855  1.271028   
2005-01-03 18:00:00  1.070830e+06   2.342437   2.432369 -0.971249  0.553060   

                            sp.13  ...      t2m.13     stl2.13     stl3.13  \
datetime                           ...                                       
2005-01-02 18:00:00  99846.319914  ...  280.473098  281.042026  281.462478   
2005-01-03 00:00:00  99917.733093  ...  278.286616  280.747406  281.486541   
2005-01-03 06:00:00  99764.378681  ...  277.206490  280.114863  281.487095   
2005-01-03 12:00:00  99672.670459  ...  280.926600  279.991138  281.472435   
2005-01-03 18:00:00  99372.811211  ...  277.363875  280.576898  281.473265   

                      iews.13   inss.13     stl4.13    fsr.13   flsr.13  \
datetime                                                                  
2005-01-02 18:00:00 -0.057958 -0.138650  284.684755  0.404731 -5.927092   
2005-01-03 00:00:00 -0.103576 -0.083050  284.667948  0.404920 -5.913881   
2005-01-03 06:00:00 -0.165721 -0.036241  284.651914  0.405704 -5.908272   
2005-01-03 12:00:00 -0.275550  0.098192  284.636266  0.403967 -5.961995   
2005-01-03 18:00:00 -0.056553  0.041844  284.620232  0.403808 -5.987860   

                      u100.13   v100.13  
datetime                                 
2005-01-02 18:00:00 -1.780562 -4.443617  
2005-01-03 00:00:00 -3.743344 -3.129469  
2005-01-03 06:00:00 -5.097203 -1.157748  
2005-01-03 12:00:00 -4.500835  1.502478  
2005-01-03 18:00:00 -3.392324  2.131114  

[5 rows x 23 columns]

Como se puede observar, hemos pasado de operar con 551 columnas, a tener sólo 23 columnas en nuestro dataset.
En último lugar, es necesario modificar la variable 'datetime' y convertir los datos de la primera columna a formato fecha (pd.to_datetime()) para facilitar su manejo y visualización. Podemos comprobar el resultado de esta modificación en los diagramas de la siguiente sección. 

In [85]:
wind_ava['datetime'] = pd.to_datetime(wind_ava['datetime'])
wind_ava.plot(figsize = (15,5))

KeyError: 'datetime'

## Descripción general de los datos

In [ ]:
wind_ava

In [ ]:
df_shape = wind_ava.shape
df_shape

Nuestro dataset tiene 23 características y 4748 instancias.

In [ ]:
df_description = wind_ava.describe()
df_description

Con esta información resumida sobre los datos disponibles podemos averiguar la media, el mínimo y el máximo y la desviación típica de cada variable, entre otras cosas. 

En primer lugar, podemos observar que todas nuestras variables son numéricas. 

También observamos que ningún valor de desviación estándar o desviación típica (std) es igual a 0, por lo que ninguna de las variables de nos interesan (las del sector 13 del mapa) son constantes, y por lo tanto no es necesario modificar el dataset más allá de lo relaizado hasta ahora.



# Preprocesado de datos
En primer lugar, queremos identificar y eliminar los datos faltantes en nuestro dataset, es decir, aquellos que son *NA*.
Para ello identificamos el número de datos faltantes por cada característica.

In [ ]:
wind_ava.isna().sum()

Como podemos ver, no hay ningún valor perdido en nuestros datos, por lo que no necesitamos eliminar ninguna característica. Lo habríamos hecho en el caso de que, para alguna característica, más del 80% de los datos fueran faltantes.

A continuación comprobamos que la desviación típica de todas las variables sea diferente de 0, lo que supondría que todas son significativas y aportan valor al modelo. En caso de que este dato tomara valor nulo supondría que esa variable es constante, y posiblemente se valoraría su eliminación del conjunto de datos.

In [ ]:
wind_ava.std()

## Visualización de los datos
Una vez se ha preprocesado y limpiado el conjunto de los datos, se procede a la creación de diferentes diagramas para recolectar información sobre la distirbución de los datos.
Con un diagrama de barras se pueden apreciar las distribuciones de energía recogidas en cada instante de tiempo (representados como el índice del conjunto de datos).

In [ ]:
wind_ava['energy'].plot(rot=20)

Para poder identificar información sobre los outliers utilizamos un diagrama de caja o boxplot.

In [ ]:
wind_ava['energy'].plot.box()

Se puede observar que hay bastantes outliers de energía generada. A pesar de que la mediana se encuentra alrededor de 500, y el tercer cuartil alrededor de 1000, existen bastantes valores fuera del rango intercuartílico, y todos ellos superando el valor de energía generada de 2500, mientras que por debajo del primer cuartil prácticamente no se percibe la existencia de outliers.

También se puede observar la frecuencia de energía en los diferentes rangos temporales (calulados con el números de divisiones *bins*) mediante un histograma.

In [ ]:
wind_ava['energy'].plot.hist(bins=25)

Se puede observar que la distribución se acumula a la izquierda de los valores temporales. Para solucionar un posible sesgo, se puede convertir la variable *energy* a logarítmica de forma que en el gráfico se aprecie como una distribución normal, en lugar de la actual que asemeja una distribución logarítmica.
Para realizar esta operación, utilizaremos la librería NumPy. Para ello inicialmente importamos la librería mencionada como se muestra a continuación:

In [ ]:
import numpy as np

In [ ]:
wind_ava['log_energy'] = np.log(wind_ava['energy'])

wind_ava['log_energy'].plot.hist(bins=40)

En este caso, el resultado no es especialmente relevante, ya que la operación convierte un sesgo a la izquierda en un sesgo a la derecha, de forma que el gráfico resultante no asemeja prácticamente una distirbución normal.

# Método de evaluación Outer & Inner
Tras haber estudiado y limpiado el conjunto de datos, se explica la estrategia que se llevará a cabo para evaluar el modelo (*outer*) y para definir el ajuste de hiperparámetros (*inner*).

*`Outer`* 
Para estimar el rendimiento esperado del modelo con datos futuros (para los que no ha sido entrenado), separamos una parte de los datos para evaluar (partición test) el modelo resultante del inner.

*`Inner`*
Emplearemos Grid Search con Cross Validation y Time Series Splits.

Para evlauar el rendimiento de los diferentes modelos se deben escoger una serie de métricas
RMSE y MAE y dummy regressor para escalar los datos y en base a ese dummy calcular RSE y RAE.
Para hacer el modelo dummy se le quita la media al real
ESto nos permite observar el error en términos absoluto y relativo
para cross validation MSE y de cara a la evaluación del modelo final tomaríamos todas esta medidas.

In [ ]:
#División de conjuntos de evaluación outer. 
#Dividimos en train y test aprovechando una falta de datos para el año 2009.
train= wind_ava[wind_ava['datetime'].dt.year<2009]
X_train= train.drop(columns='energy')
y_train= train['energy']
print("train shape is:", train.shape,"\n X_train shape:",X_train.shape, "\n y_train shape: ", y_train.shape)

test= wind_ava[wind_ava['datetime'].dt.year>2008]
X_test= test.drop(columns='energy')
y_test= test['energy']
print("test shape is:", test.shape,"\n X_test shape:",X_test.shape, "\n y_train shape: ", y_test.shape)
print(train.shape, test.shape, wind_ava.shape)

# Selección de Scaler
En esta sección se emplea KNN y CV para encontrar el método de escalado que minimice el error entre las tres opciones que ofrece sklearn: MinMax, Standarization y Robust. 
Para ello, empleamos un sencillo crossvalidation con un split adecuado para series temporales y medimos el error empleando la métrica escogida (MSE).

In [91]:
#1) Preparación de datos -usaremos el subset train. 
scalerCV= TimeSeriesSplit(n_splits=5)
scaler_eval= {}
#2)Evaluación de scalers
#StandardScaler
pipeline_std= sklearn.pipeline.Pipeline([ ('standardScaler', preprocessing.StandardScaler()),
                                 ('KNN',neighbors.KNeighborsRegressor()) ])
score_std= cross_val_score(pipeline_std, X_train, y_train,cv= scalerCV, score=metrics.mean_squared_error)
scaler_eval["StandardScaler"] = score_std.mean()
print(scaler_eval["StandardScaler"])
#MinMaxScaler

#RobustScaler


AttributeError: module 'sklearn' has no attribute 'pipeline'